### ISSUE TITLE
provide the correct way to go from addr1 format to pubkeyhash

### ISSUE THREAD

**Issue by Benjmhart** [Plutonomicon/cardano-browser-tx#9]
nami wallet provides addr1-formatted addresses

in previous attempts to integrate with Nami, we had difficulty getting the _correct_ pubkeyhash out of this address.

we may need to get the right library functions to do this correctly.

**Comment by mikekeke** [comment]
A I understand, Nami gives address that built from `Credential` and `StakingCredential`: https://staging.plutus.iohkdev.io/doc/haddock/plutus-ledger-api/html/src/Plutus.V1.Ledger.Address.html#Address

`Credential` could be `PubKeyHash` or `ValidatorHash`: https://staging.plutus.iohkdev.io/doc/haddock/plutus-ledger-api/html/src/Plutus.V1.Ledger.Credential.html#Credential

So, when we parse `Address`, we can get `PubKeyHash` from it, but we can not build back same `Address` from `PubKeyHash` without `StakingCredential`, it it wasn't `Nothing`. So if we are using address with `StakingCredential` like Nami provides, we can't build that address only from `PubKeyHash` .

**Comment by adamczykm** [comment]
Can't yet assign myself.

**Comment by nrutledge** [comment]
Yeah, there is no issue going from addr1 format to pubKeyHash. Going from pubKeyHash back to addr1 is the issue due to loss of StakingCredential as Misha stated. Solution seems to be to always store the full address as opposed to just the PubKeyHash.

**Comment by adamczykm** [comment]
### **Getting PubKeyHash**

There exists a function in nami-wallet api that allows to extract a PKH from BaseAddress:

```jsx
export const extractKeyHash = async (address) => {
  await Loader.load();
  //TODO: implement for various address types
  if (!(await isValidAddressBytes(Buffer.from(address, 'hex'))))
    throw DataSignError.InvalidFormat;
  try {
    const baseAddr = Loader.Cardano.BaseAddress.from_address(
      Loader.Cardano.Address.from_bytes(Buffer.from(address, 'hex'))
    );
    return baseAddr.payment_cred().to_keyhash().to_bech32('hbas_');
  } catch (e) {}
  try {
    const rewardAddr = Loader.Cardano.RewardAddress.from_address(
      Loader.Cardano.Address.from_bytes(Buffer.from(address, 'hex'))
    );
    return rewardAddr.payment_cred().to_keyhash().to_bech32('hrew_');
  } catch (e) {}
  throw DataSignError.AddressNotPK;
};
```

### Getting BaseAddress

Nami creates its keys and addresses using code snippets provided in  **cardano-serialization-lib.**

Following its documentation it seems that if we want to have an address with staking ability we do need to provide staking credentials.

*from  [https://github.com/Emurgo/cardano-serialization-lib/blob/master/doc/getting-started/generating-keys.md](https://github.com/Emurgo/cardano-serialization-lib/blob/master/doc/getting-started/generating-keys.md):*

---

Once we have reached the desired derivation path, we must convert the `BIP32PrivateKey` or `BIP32PublicKey` to a `PrivateKey` or `PublicKey` by calling `.to_raw_key()` on them with the exception of Byron addresses.
For example, to create an address using the `utxoPubKey` and `stakeKey` in the first example, we can do:

```jsx
// base address with staking key
const baseAddr = CardanoWasm.BaseAddress.new(
  CardanoWasm.NetworkInfo.mainnet().network_id(),
  CardanoWasm.StakeCredential.from_keyhash(utxoPubKey.to_raw_key().hash()),
  CardanoWasm.StakeCredential.from_keyhash(stakeKey.to_raw_key().hash()),
);

// enterprise address without staking ability, for use by exchanges/etc
const enterpriseAddr = CardanoWasm.EnterpriseAddress.new(
  CardanoWasm.NetworkInfo.mainnet().network_id(),
  CardanoWasm.StakeCredential.from_keyhash(utxoPubKey.to_raw_key().hash())
);

// pointer address - similar to Base address but can be shorter, see formal spec for explanation
const ptrAddr = CardanoWasm.PointerAddress.new(
  CardanoWasm.NetworkInfo.mainnet().network_id(),
  CardanoWasm.StakeCredential.from_keyhash(utxoPubKey.to_raw_key().hash()),
  CardanoWasm.Pointer.new(
    100, // slot
    2,   // tx index in slot
    0    // cert indiex in tx
  )
);
```

So it seems that we can have following functions 

```jsx
newBaseAddress :: NetworkId -> PubKeyHash -> StakeKeyHash -> BaseAddress
addrToPubKeyHash :: BaseAddress -> PubKeyHash
```

**Comment by adamczykm** [comment]
To resolve the 'issue' I will implement these functions.

**Comment by ngua** [comment]
Closed by #31
